File: src/services/invoicing_service.py
from sqlalchemy.orm import Session
from src.models.invoice import Invoice
from src.models.billing import Subscription, Plan
from src.models.transaction import Transaction
from datetime import datetime, timedelta
from loguru import logger
from typing import List, Dict, Any
import uuid

class InvoicingService:
    """
    Gerador de Faturas e Dashboards Financeiros.
    Replica a lógica de 'AccountingEngine' do .NET.
    """
    @staticmethod
    def get_user_dashboard(db: Session, tenant_id: str) -> Dict[str, Any]:
        """
        Retorna visão geral financeira para o Tenant.
        """
        # 1. Busca Assinatura Ativa
        sub = db.query(Subscription).filter(Subscription.tenant_id == tenant_id).first()
        plan_name = sub.plan.name if sub else "Nenhum"
        next_billing = sub.expires_at if sub else None
        
        # 2. Histórico de Pagamentos recentes
        transactions = db.query(Transaction).filter(
            Transaction.tenant_id == tenant_id,
            Transaction.status == "approved"
        ).order_by(Transaction.created_at.desc()).limit(5).all()
        
        # 3. Faturas em Aberto
        open_invoices = db.query(Invoice).filter(
            Invoice.tenant_id == tenant_id,
            Invoice.status == "open"
        ).all()
        
        return {
            "current_plan": plan_name,
            "next_billing_date": next_billing,
            "total_spent": sum(t.amount for t in transactions),
            "recent_payments": [
                {"id": t.id, "date": t.created_at, "amount": t.amount, "status": t.status}
                for t in transactions
            ],
            "invoices": [
                {"id": i.id, "number": i.invoice_number, "status": i.status, "amount": i.amount}
                for i in open_invoices
            ]
        }

    @staticmethod
    def generate_monthly_invoice(db: Session, tenant_id: str) -> Invoice:
        """Gera uma nova fatura (Draft) para o ciclo atual (Sprint 34)."""
        sub = db.query(Subscription).filter(Subscription.tenant_id == tenant_id).first()
        if not sub: raise Exception("Tenant sem plano")
        
        invoice = Invoice(
            tenant_id=tenant_id,
            invoice_number=f"INV-{str(uuid.uuid4())[:8].upper()}",
            period_start=datetime.utcnow(),
            period_end=datetime.utcnow() + timedelta(days=30),
            amount=sub.plan.price,
            plan_name=sub.plan.name,
            status="open"
        )
        db.add(invoice)
        db.commit()
        db.refresh(invoice)
        logger.info(f"📄 Fatura gerada: {invoice.invoice_number} para Tenant {tenant_id}")
        return invoice

File: node-version/src/services/billing/invoicingService.js
const { Invoice, Subscription, Transaction } = require('../../models/sql/models');
const { v4: uuidv4 } = require('uuid');
const logger = require('../../utils/logger');
const moment = require('moment');

class InvoicingService {

  static async getUserDashboard(tenantId) {
    const sub = await Subscription.findOne({
      where: { tenant_id: tenantId },
      include: ['Plan']
    });

    const planName = sub && sub.Plan ? sub.Plan.name : 'Nenhum';
    const nextBilling = sub ? sub.expires_at : null;

    const transactions = await Transaction.findAll({
      where: { tenant_id: tenantId, status: 'approved' },
      order: [['created_at', 'DESC']],
      limit: 5
    });

    const openInvoices = await Invoice.findAll({
      where: { tenant_id: tenantId, status: 'open' }
    });

    const totalSpent = transactions.reduce((acc, t) => acc + t.amount, 0);

    return {
      current_plan: planName,
      next_billing_date: nextBilling,
      total_spent: totalSpent,
      recent_payments: transactions.map(t => ({
        id: t.id,
        date: t.created_at,
        amount: t.amount,
        status: t.status
      })),
      invoices: openInvoices.map(i => ({
        id: i.id,
        number: i.invoice_number,
        status: i.status,
        amount: i.amount
      }))
    };
  }

  static async generateMonthlyInvoice(tenantId) {
    const sub = await Subscription.findOne({
      where: { tenant_id: tenantId },
      include: ['Plan']
    });

    if (!sub) throw new Error('Tenant sem plano');

    const invoice = await Invoice.create({
      tenant_id: tenantId,
      invoice_number: `INV-${uuidv4().substring(0, 8).toUpperCase()}`,
      period_start: new Date(),
      period_end: moment().add(30, 'days').toDate(),
      amount: sub.Plan.price,
      plan_name: sub.Plan.name,
      status: 'open'
    });

    logger.info(`📄 Fatura gerada: ${invoice.invoice_number} para Tenant ${tenantId}`);
    return invoice;
  }
}

module.exports = InvoicingService;
